### PR TITLE
Add query building API for the `$densify` aggregation pipeline stage

### PR DESCRIPTION
**Pull request by stIncMale**
The new API:

- Aggregates.densify for [`$densify`](https://www.mongodb.com/docs/v6.0/reference/operator/aggregation/densify/)

The server-side syntax document is [here](https://docs.google.com/document/d/1pwIBc5XzbZkMnZH-FAoVYf2OOUzQM2VT8wbtoC4FMTg/edit#heading=h.sshyb2wdxr2w).

You may see how the API usage looks like, or play with it in

- `AggregatesFunctionalSpecification.'$densify'`

JAVA-4350

### CODE REVIEW

File: driver-core/src/main/com/mongodb/client/model/MongoTimeUnit.java [hunk 1]
```diff
@@ -68,7 +73,13 @@ public enum MongoTimeUnit {
...
     }
 
-    String value() {
+    /**
+     * Returns a {@link String} representation of the unit, which may be useful when using methods like
+     * {@link Windows#of(Bson)}, {@link DensifyRange#of(Bson)}.
+     *
+     * @return A {@link String} representation of the unit.
+     */
+    public String value() {
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/client/model/MongoTimeUnit.java:82]: Had to make it `public` to be able to access it from the new `com.mongodb.client.model.densify` package.

**Comment by stIncMale** [comment]
Converted to draft to express step as a separate interface. This will simplify things.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/client/model/densify/DensifySpecificationBase.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+     *
+     * @param range The range.
+     * @return A new {@link DensifySpecification}.
+     */
+    DensifySpecification range(DensifyRange range);
+
+    /**
+     * Creates a new base for {@link DensifySpecification} with the specified {@code range}.
+     *
+     * @param range The range.
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifySpecificationBase.java:43]: Currently neither `range(DensifyRange)` nor `range(PartitionByFieldDensifyRange)` contains `@see DensifyRange`/`@see PartitionByFieldDensifyRange`. However, that would be weird given that links to those types are already in the documentation of the methods, as the types are used by parameters. And docs for both `DensifyRange` and `PartitionByFieldDensifyRange` link to factory methods that may be used to create them.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+     * @param step The positive step.
+     * @return The requested {@link DensifyRange}.
+     */
+    static NumberDensifyRange numberRangeWithStep(final Number l, final Number u, final Number step) {
+        notNull("l", l);
+        notNull("u", u);
+        if (l instanceof Comparable && l.getClass().isAssignableFrom(u.getClass())) {
+            @SuppressWarnings("unchecked")
+            Comparable<Number> comparableL = (Comparable<Number>) l;
+            // a half open interval with equal bounds is an empty interval, and the server allows specifying it
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java:89]: And we don't prevent it, we allow it by including `=` in the `<=` check. I left a comment because I found this surprising, and wanted to explain why the check allows specifying an empty interval.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifySpecification.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * A specification of how to densify a sequence of documents.
+ *
+ * @see Aggregates#densify(DensifySpecification)
+ * @mongodb.server.release 5.1
+ * @since 4.7
+ */
+@Evolving
+public interface DensifySpecification extends PartitionByFieldDensifySpecificationBase, Bson {
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifySpecification.java:33]: The reason I did not call it `Densify`, which I like more than `DensifySpecification`, is that objects of this type do not represent a complete pipeline stage. The stage is created by `Aggregates.densify(DensifySpecification)`, and I wanted to avoid confusion that may arise by seeing `Densify` and assuming it represents a stage.

However, if with my answer you still think it is better to name this type `Densify`, I will gladly do so.

P.S. I am not sure if "densify" is made up, because both [Google](https://www.google.com/search?q=densify&ei=oLGrYoDcCMS-kPIPmdeiwA8&ved=0ahUKEwiArOe4hbP4AhVEH0QIHZmrCPgQ4dUDCA4&uact=5&oq=densify&gs_lcp=Cgdnd3Mtd2l6EAMyEQguELEDEIMBEMcBENEDEJECMgUIABCRAjIFCAAQkQIyBQgAEIAEMgUIABCABDIFCAAQgAQyCwguEIAEEMcBEK8BMgUIABCABDIFCAAQgAQyBQgAEIAEOgcIABBHELADSgQIQRgASgQIRhgAULUEWLUEYJsIaAJwAXgAgAFqiAFqkgEDMC4xmAEAoAEByAEIwAEB&sclient=gws-wiz) and dictionaries ([here](https://www.merriam-webster.com/dictionary/densify), [here](https://www.collinsdictionary.com/dictionary/english/densify)) seem to consider it a real word.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+     * Returns a {@link DensifyRange} that represents an interval with the smallest BSON {@link BsonType#DATE_TIME Date} value
+     * of the {@linkplain DensifySpecification#densifyField(String) field}
+     * in the sequence of documents being its lower bound, and the largest value being the upper bound.
+     *
+     * @param step The positive step.
+     * @param unit The unit in which the {@code step} is specified.
+     * @return The requested {@link DensifyRange}.
+     */
+    static DateDensifyRange dateRangeFullWithStep(final long step, final MongoTimeUnit unit) {
+        isTrueArgument("step must be positive", step > 0);
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java]: Usually, if checking is simple, I prefer to do in in the driver. I removed the checking as I don't care much about them. However, I wonder if you have a reason beyond "server does the check anyway" for this particular case? If not, should we then stop ever doing runtime checks for MQL, and should I remove other runtime checks in this PR?

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java:89]: oh, i see.  But do we need the check at all?  If the server reports an error when `l > u` then we don't need to do so in the driver.  This is consistent with general driver practice of allowing the server to enforce constraints.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifySpecificationBase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifySpecificationBase.java:43]: ok, so it's a double hop.

File: driver-core/src/main/com/mongodb/client/model/MongoTimeUnit.java (hunk 1 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/client/model/MongoTimeUnit.java:82]: Ack

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java]: It's one of the mantras: https://github.com/mongodb/specifications#where-possible-depend-on-server-to-return-errors.  In general, I don't think we need to do runtime checks.  The purpose of the builders is to enforce compile-time constraints and aid in discovery.  Runtime errors don't help the user much, as it's not likely people will be be unit testing their builder usage, and we have to write and maintain tests asserting our client-side validation.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java]: Done.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java]: Done.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifySpecification.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifySpecification.java:33]: > builders for those stages called Filters and Projections

They are called `Filter`**s** and `Projection`**s** because those classes do not define a type used by either driver or a user, instead, they are simply compartments for factory methods. The same is true for `Windows` and `WindowedComputations`. Notice how the actual types they build are called `Window` and `WindowedComputation` respectively.

The `$search` query building API went further: why having a separate class that is just a holder for factory methods, when those factory methods may be contained by the interface corresponding to the type which instances are constructed by those factory methods? (the ability to define `static` methods in interfaces was added to Java because it [_"makes it easier for you to organize helper methods in your libraries; you can keep static methods specific to an interface in the same interface rather than in a separate class"_](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html#static)).

Given the above, `DensifySpecification` may be renamed to `Densify` or `Densification`, but `Densifications` (plural) does not appear to be the right way to name this type.

> I think I prefer Densifications

Is the proposed name `Densification` instead of `Densify` intended to reduce the probability of user thinking that an instance of `Densification` is a complete `$densify` stage? If yes, then using this name may help in this case, but we may not always come of with similar names for other future stages. For example, doing it for `$out` or `bucket` does not seem possible. This is why I used `DensifySpecification` - this naming approach will fork for any stage. However, the approach of calling it `Densify` will also work for for any stage.

**Comment by stIncMale** [comment]
Converted to draft until there is a decision on whether we should use a different approach proposed by @katcharov.

**Comment by stIncMale** [comment]
Moved back to "Ready for review".

**Comment by stIncMale** [comment]
The PR was refactored. A single pipeline stage specification is no longer used. Instead

- the required fields are represented as params of `Aggregates.densify` methods;
- the API uses an object that represents optional fields (similarly to `Aggregates.bucket`, `Aggregates.search`; the API of this object has the same style as the API of `SearchOptions`).

File: driver-core/src/main/com/mongodb/client/model/densify/PartitionByFieldDensifySpecificationBase.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+@Evolving
+public interface PartitionByFieldDensifySpecificationBase {
+    /**
+     * Creates a new {@link DensifySpecification} with the specified {@code field} to partition by.
+     *
+     * @param field The field to partition by.
+     * @return A new {@link DensifySpecification}.
+     * @mongodb.driver.manual core/document/#dot-notation Dot notation
+     */
+    DensifySpecification partitionByField(String field);
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/PartitionByFieldDensifySpecificationBase.java]: Resolving because refactoring made the question irrelevant.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifySpecification.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifySpecification.java]: Resolving because refactoring made the question irrelevant.

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ * @since 4.7
+ */
+@Evolving
+public interface DensifyOptions extends Bson {
+    /**
+     * Creates a new {@link DensifyOptions} with the specified {@code fields} to partition by.
+     *
+     * @param fields The fields to partition by.
+     * @return A new {@link DensifyOptions}.
+     * @mongodb.driver.manual core/document/#dot-notation Dot notation
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java:41]: Is this the right link its not obvious how its related

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    default DensifyOptions partitionByFields(@Nullable final String... fields) {
+        return partitionByFields(fields == null ? emptyList() : asList(fields));
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java:44]: This removal behaviour of null / an empty list is undocumented

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+     * @param step The step.
+     * @return The requested {@link DensifyRange}.
+     */
+    static NumberDensifyRange partitionRangeWithStep(final Number step) {
+        return new DensifyConstructibleBson(new Document("bounds", "partition")
+                .append("step", notNull("step", step)));
+    }
+
+    /**
+     * Returns a {@link DensifyRange} that represents a single interval [l; u).
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java:78]: is `[l; u)` syntax for lower bound inclusive to upper bound?

**Review by rozza — CHANGES_REQUESTED** [review]
Looks good - just a couple of minor comments

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java (hunk 7 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java:41]: This is a documentation for `partitionByFields`, which accepts fields, and this link suggests that the dot-notation is allowed (in some places we allow the dot-notation for fields, and in some we don't).

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java (hunk 8 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java:45]: I updated the documentation to clarify what it means when no fields to partition by are specified. However, there is no removal behavior that can be documented here, because we don't remove anything: we always create a new `DensifyOptions` according to the specified fields (whether there are none or some).

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java (hunk 9 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java:78]: It's not, it's the standard interval notation: https://en.wikipedia.org/wiki/Interval_(mathematics). The documentation for `DensifyRange` also describes the interval in plain English: _"a half-closed interval of values with the lower bound being inclusive"_.

Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java:78]: Thanks, good to know - I just hadn't seen it with a semi-colon, just a comma. 

So its similar to the [maven ranges](https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#MAVEN402).

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java (hunk 7 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java:42]: Perhaps mention dot notation is supported in the docs above?

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java (hunk 8 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyOptions.java:45]: Thanks, I think that's clarified the behaviour of `opts = myOptions.partitionByFields('a', 'b').partitionFields(null)`

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java (hunk 9 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java]: No worries - I learnt something new with the wiki page for intervals :)

**Review by katcharov — APPROVED** [review]
Just one minor comment.

lgtm

File: driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    /**
+     * Returns a {@link DensifyRange} that represents a single interval [l, u).
+     *
+     * @param l The lower bound.
+     * @param u The upper bound.
+     * @param step The step.
+     * @param unit The unit in which the {@code step} is specified.
+     * @return The requested {@link DensifyRange}.
+     */
+    static DateDensifyRange rangeWithStep(final Instant l, final Instant u, final long step, final MongoTimeUnit unit) {
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java:135]: Returning a more specific type has the following benefits:

- We may introduce type-specific methods ("setters"), thus expressing that an optional field is present on `DateDensifyRange` and not present on `NumberDensifyRange`.
- We may accept a more specific type as a parameter in some methods, thus guiding a user towards using the API correctly, instead of expressing the type restrictions in the docs. For example, we can accept `DateDensifyRange`, instead of accepting `DensifyRange` and additionally documenting that only objects created as date ranges are allowed to be used.

Both of the above benefits were used in the `$search`/`$searchMeta` API. In this PR I am just following the idea in case it becomes handy in the future (we won't be able to return a more specific type in the future without breaking binary compatibility).
